File: dec test/xtest/app/src/main/java/diff/rednaga/AXMLPrinter.java

package diff.rednaga;

import android.content.res.AXMLResource;

import java.io.*;
import java.util.Properties;

public class AXMLPrinter {

    private static String VERSION;

    static {
        InputStream templateStream = AXMLPrinter.class.getClassLoader().getResourceAsStream("axmlprinter.properties");
        if (templateStream != null) {
            Properties properties = new Properties();
            String version = "(unknown version)";
            try {
                properties.load(templateStream);
                version = properties.getProperty("application.version");
            } catch (IOException ex) {
                System.err.println("Unable to find version number!");
            }
            VERSION = version;
        } else {
            VERSION = "[unknown version - no properties found]";
        }

    }

    public static void main(String[] arguments) throws IOException {
        if (arguments.length < 1) {
            System.out.println("Usage: AXMLPrinter <binary xml file>");
            return;
        }

        if (arguments[0].equalsIgnoreCase("-v") || arguments[0].equalsIgnoreCase("-version")) {
            //System.out.printf("axmlprinter %s (http://github.com/rednaga/axmlprinter2)\n", VERSION);
            //System.out.printf("Copyright (C) 2015-2025 Red Naga - Tim 'diff' Strazzere (diff@protonmail.com)\n");
            return;
        }

        FileInputStream fileInputStream = null;
        FileOutputStream fileOutputStream = null;
        try {
            AXMLResource axmlResource = new AXMLResource();
            fileInputStream = new FileInputStream(arguments[0]);
            axmlResource.read(fileInputStream);

            axmlResource.print();

            if (arguments.length > 1) {
                File file = new File(arguments[1]);
                fileOutputStream = new FileOutputStream(file);
                axmlResource.write(fileOutputStream);
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (fileInputStream != null) {
                fileInputStream.close();
            }

            if (fileOutputStream != null) {
                fileOutputStream.close();
            }
        }
    }

    /*
     * Avoid anyone accidentally (purposefully?) Instantiating this class
     */
    private AXMLPrinter() {

    }
}

File: dec test/xtest/app/src/main/java/com/container/utils/AXMLPrinter.kt
// File: app/src/main/java/com/container/utils/AXMLPrinter.kt
package com.container.utils

import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.util.Properties

class AXMLPrinter private constructor() {

    companion object {
        private const val VERSION_PROPERTIES = "axmlprinter.properties"
        private var version: String = "(unknown version)"

        init {
            loadVersionInfo()
        }

        private fun loadVersionInfo() {
            AXMLPrinter::class.java.classLoader?.getResourceAsStream(VERSION_PROPERTIES)?.use { stream ->
                Properties().apply {
                    load(stream)
                    version = getProperty("application.version", version)
                }
            } ?: run {
                println("Unable to find version properties!")
            }
        }

        fun printVersion() {
            println("axmlprinter $version (http://github.com/rednaga/axmlprinter2)")
            println("Copyright (C) 2015-2025 Red Naga - Tim 'diff' Strazzere (diff@protonmail.com)")
        }

        fun parseManifest(axmlFile: File): PackageInfo {
            require(axmlFile.exists()) { "Binary XML file not found: ${axmlFile.absolutePath}" }

            return try {
                FileInputStream(axmlFile).use { input ->
                    AXMLResource().apply {
                        read(input)
                        print() // Optional: Print to console for debugging
                    }.toPackageInfo()
                }
            } catch (e: Exception) {
                throw IOException("Failed to parse AXML file", e)
            }
        }

        fun writeManifest(axmlFile: File, outputFile: File) {
            require(axmlFile.exists()) { "Binary XML file not found: ${axmlFile.absolutePath}" }

            try {
                FileInputStream(axmlFile).use { input ->
                    FileOutputStream(outputFile).use { output ->
                        AXMLResource().apply {
                            read(input)
                            write(output)
                        }
                    }
                }
            } catch (e: Exception) {
                throw IOException("Failed to write AXML file", e)
            }
        }
    }
}